lib/alist.3
.\" -*- nroff -*-
.\" $Id: alist.3,v 1.1 2003/04/24 15:11:58 rrt Exp $
.Dd February 12, 2001
.Os
.Dt ALIST 3
.Sh NAME
.Nm alist
.Nd Doubly-linked lists
.Sh SYNOPSIS
.Fd #include <alist.h>
.Bd -literal

typedef /* ... */ alist;

alist	alist_new(void);
alist	alist_copy(alist al);
void	alist_delete(alist al);
void	alist_clear(alist al);
int	alist_count(alist al);
int	alist_isempty(alist al);
void *	alist_first(alist al);
void *	alist_last(alist al);
void *	alist_prev(alist al);
void *	alist_next(alist al);
void *	alist_current(alist al);
int	alist_current_idx(alist al);
void	alist_insert(alist al, unsigned int i, void *p);
void	alist_prepend(alist al, void *p);
void	alist_append(alist al, void *p);
int	alist_remove(alist al);
int	alist_remove_ptr(alist al, void *p);
int	alist_remove_idx(alist al, unsigned int i);
void *	alist_take(alist al);
void *	alist_take_idx(alist al, unsigned int i);
void	alist_sort(alist al, int (*cmp)(const void *p1, const void *p2));
void *	alist_at(alist al, unsigned int i);
int	alist_find(alist al, void *p);
.Ed
.Sh DESCRIPTION
The
.Nm
library provides a high-level interface to doubly-linked lists.
.Pp
The library provides one type:
.Fa alist .
.Pp
The
.Fn alist_new
function allocates a new empty list.
.Pp
The
.Fn alist_copy
function allocates a new list and fills it with the items of the argument
list
.Fa al .
.Pp
The
.Fn alist_delete
function 
deallocates the argument list
.Fa al .
.Pp
The
.Fn alist_clear
function removes all the items from the argument list
.Fa al .
.Pp
The
.Fn alist_count
function returns the number of items contained in the argument list
.Fa al .
.Pp
The
.Fn alist_isempty
function returns a zero value if the argument list
.Fa al
is empty, non-zero otherwise.
.Pp
The
.Fn alist_first
function returns the first item of the argument list
.Fa al .
.Pp
The
.Fn alist_last
function returns the last item of the argument list
.Fa al .
.Pp
The
.Fn alist_prev
function returns the previous item of the argument list
.Fa al .
.Pp
The
.Fn alist_next
function returns the next item of the argument list
.Fa al .
.Pp
The
.Fn alist_current
function returns the current item of the argument list
.Fa al .
.Pp
The
.Fn alist_current_idx
function returns the current item index of the argument list
.Fa al .
The first item has a zero index, while the latest item has
.Fa alist_count(al)-1
index.
.Pp
The
.Fn alist_insert
function inserts the
.Fa p
item at position
.Fa i .
Previously existing items at position
.Fa i 
will be moved after the inserted item.
.Pp
The
.Fn alist_append
function appends the
.Fa p
item to the argument list
.Fa al .
.Pp
The
.Fn alist_prepend
function prepends the
.Fa p
item to the argument list
.Fa al .
.Pp
The
.Fn alist_remove
function removes the current item from the argument list
.Fa al .
The function will return
a zero value if the operation was successful, -1 otherwise.
.Pp
The
.Fn alist_remove_ptr
function removes the argument item
.Fa p
from the argument list
.Fa al .
The function will return
a zero value if the operation was successful, -1 otherwise.
.Pp
The
.Fn alist_remove_idx
function removes the item at index
.Fa i
from the argument list
.Fa al .
The function will return
a zero value if the operation was successful, -1 otherwise.
.Pp
The
.Fn alist_take
function removes the current item from the argument list
.Fa al .
The function will return the removed item.
.Pp
The
.Fn alist_take_idx
function removes the item at index
.Fa i
from the argument list
.Fa al .
The function will return the removed item.
.Pp
The
.Fn alist_sort
functions sorts the argument list
.Fa al
using the comparison function pointed by the argument
.Fa cmp .
.Pp
The contents of the list are sorted in ascending order
according to a comparison function pointed to by
.Fa cmp ,
which is called with two arguments that point to the
objects being compared.
The comparison function must return an integer less than,
equal to, or greater than zero if the first argument is
considered to be respectively less than, equal to, or
greater than the second. If two members compare as equal,
their order in the sorted list is undefined.
.Pp
The
.Fn alist_at
function returns the item at index
.Fa i
contained in the argument list
.Fa al .
.Pp
The
.Fn alist_find
function finds the first occurrence of the argument item
.Fa p
into the argument list
.Fa al .
.Sh DEBUGGING
If you would like to debug your program, you should define the macro
.Fa ALIST_NO_MACRO_DEFS
before including the header of this library, i.e.
.Bd -literal -offset indent
#define ALIST_NO_MACRO_DEFS
#include <alist.h>
.Ed
.Pp
This prevents defining at least the following function macros that makes
code faster but debugging harder:
.Fn alist_isempty ,
.Fn alist_count ,
.Fn alist_first ,
.Fn alist_last ,
.Fn alist_prev ,
.Fn alist_next ,
.Fn alist_current ,
.Fn alist_current_idx .
Side effects (like incrementing the argument) in parameters of these macros
should be avoided.
.Sh EXAMPLES
Sorting and printing a list of words:
.Bd -literal -offset indent
int sorter(const void *p1, const void *p2)
{
	return strcmp(*(char **)p1, *(char **)p2);
}

int main(void)
{
	alist al;
	char *s;
	al = alist_new();
	alist_append(al, "long");
	alist_append(al, "int");
	alist_append(al, "double");
	alist_append(al, "char");
	alist_append(al, "float");
	alist_sort(al, sorter);
	for (s = alist_first(al); s != NULL; s = alist_next(al))
		printf("Item #%d: %s\\n",
		       alist_current_idx(al) + 1, s);
	alist_delete(al);
	return 0;
}
.Ed
.Pp
An address book:
.Bd -literal -offset indent
typedef struct person {
	char *first_name;
	char *last_name;
	char *phone;
} person;

int sorter(const void *p1, const void *p2)
{
	person *pe1 = *(person **)p1, *pe2 = *(person **)p2;
	int v;
	if ((v = strcmp(pe1->last_name, pe2->last_name)) == 0)
		v = strcmp(pe1->first_name, pe2->first_name);
	return v;
}

int main(void)
{
	alist al;
	person *p, p1, p2, p3, p4;
	al = alist_new();
	p1.first_name = "Joe";
	p1.last_name = "Green";
	p1.phone = "555-123456";
	p2.first_name = "Joe";
	p2.last_name = "Doe";
	p2.phone = "555-654321";
	p3.first_name = "Dennis";
	p3.last_name = "Ritchie";
	p3.phone = "555-314159";
	p4.first_name = "Brian";
	p4.last_name = "Kernighan";
	p4.phone = "555-271828";

	alist_append(al, &p1);
	alist_append(al, &p2);
	alist_append(al, &p3);
	alist_append(al, &p4);
	alist_sort(al, sorter);
	for (p = alist_first(al); p != NULL; p = alist_next(al))
		printf("Person #%d: %10s, %-10s Tel.: %s\\n",
		       alist_current_idx(al) + 1,
		       p->first_name, p->last_name, p->phone);
	alist_delete(al);
	return 0;
}
.Ed
.Pp
A list of lists:
.Bd -literal -offset indent
alist al1, al2, al3, al;
char *s;
al1 = alist_new();
al2 = alist_new();
al3 = alist_new();
alist_append(al2, "obj1");
alist_append(al2, "obj2");
alist_append(al2, "obj3");
alist_append(al3, "obja");
alist_append(al3, "objb");
alist_append(al3, "objc");
alist_append(al1, al2);
alist_append(al1, al3);
for (al = alist_first(al1); al != NULL; al = alist_next(al1))
	for (s = alist_first(al); s != NULL; s = alist_next(al))
		printf("String: %s\\n");
alist_delete(al1);
alist_delete(al2);
alist_delete(al3);
.Ed
.Sh AUTHORS
Sandro Sigala <sandro@sigala.it>


NEWS
Changes between Zile 1.6.1 - Zile 1.7 BETA 2

Note: 1.7 BETA 2 was produced from 1.6.1, not 1.6.2, but all the same
fixes were applied.

* Changed configure script so no syntax colouring modes are built by
  default.
* Implemented Java Mode.  Contributed by Marius Nita <marius@cs.pdx.edu>.
* Implemented C# Mode.  Contributed by Josef 'Jupp' Schugt <jupp@gmx.de>.
* Implemented Mail Mode (plus the five variables `font-mailN' and
  the variable `mail-mode-auto-fill') as suggested by Josef Schugt.
* Implemented the widely-used Emacs bindings:
    - `forward-word' (bound to `M-f'), `backward-word' (bound to `M-b');
    - `capitalize-word' (bount to `M-c');
    - `upcase-word' (bount to `M-u'), `downcase-word' (bount to `M-l');
* Implemented `describe-key-briefly' (bound to `C-h c').
* Implemented `shrink-window' (bound to `M-e').  Contributed by Marius Nita.
* Implemented `back-to-indentation' (bound to `M-m').  Contributed by
  Joseph Boyd <k963@bonbon.net>.
* Implemented `goto-char'.
* Implemented `revisions-kept' and `revisions-delete' variables as
  suggested by Josef Schugt.
* Added `-u rcfile' and `-v variable=value' command line options as
  requested by many users.
* Now multiple functions can be specified on command line with `-f func'.
* Now `quoted-insert' (bound to `C-q') accepts octal digits to
  insert particular characters.
* Cleaned up X Window terminal resize handling.  Low level SIGWINCH signal
  handling has been removed; now Zile relies only on ncurses KEY_RESIZE.
* The behavior of the `auto-font-lock' variable is changed in this version.
  Now the variable applies also to buffer mode changes.
* Fixed two old bugs thanks to Reuben Thomas <rrt@mupsych.org>.
* Patches provided by David A. Capello <dacap@users.sourceforge.net>:
    * implemented `kill-word', `backward-kill-word';
    * fixed bug in undo/redo;
    * fixed bug in highlight;
    * macros support `isearch' facility;
    * shell mode supports highlight for "cat<<EOF\n...\nEOF" stuff.  With
      this, added a new variable: `font-here-document';
    * better shell mode highlight: $identifier, and fixed the \'string\'
      problem;
    * optimized the selection highlight color;
    * added alias `sh-mode' to `shell-script-mode'.
* Integrated some excellent patches from Nicolas Duboc.
* Speed-up improvements.
* Many many other bug fixes.

Changes between Zile 1.6 - Zile 1.6.1

* Added `.m' extension to C Mode (for Objective-C files).
* Fixed window resize bug under xterm (\377 character was inserted).
* Typos and minor bug fixes.

Changes between Zile 1.5.3 - Zile 1.6

* Implemented incremental search (`isearch-forward', `isearch-backward').
* Implemented search & replace (`replace-string', `query-replace').
* Implemented shell commands (`shell-command', `shell-command-on-region').
* Implemented better handling of unexpected errors, with automatic file
  saving on Zile crash.

Changes between Zile 1.5.2 - Zile 1.5.3

* Fixed wrong behavior in novice mode (Font Lock automatically set).
* Fixed a few core dumps.

Changes between Zile 1.5.1 - Zile 1.5.2

* Implemented Shell-script Mode.
* Fixed Solaris compatibility bugs.
* Fixed bugs in C/C++ Font Lock parsing.

Changes between Zile 1.5 - Zile 1.5.1

* Fixed Minibuf filename completion bug.

Changes between Zile 1.4 - Zile 1.5

* Added `switch-to-correlated-buffer' (bound to `F11').
* Added `skip-splash-screen' variable.
* Extented Mini Help functionality; now there are three help pages.
* Reorganized partially the `F1'..`F12' keys.
* Added a reference card (see `etc/refcard.tex' and `etc/refcard.ps').
* User-invisible changes: introduced three new general-purpose libraries:
	- lib/alist.c	Doubly-linked lists
	- lib/astr.c	Dynamically allocated strings
	- lib/htable.c	Hash table functions (was hash.c)
  The new libraries are documented in their man pages (see lib/*.3).
  Starting from this version, the internal structures will be rewritten
  using these libraries.  This will make the code easier to maintain,
  and hopefully fixes some hidden bugs.
* Many bug fixes.

Changes between Zile 1.3 - Zile 1.4

* Added support for `F1'..`F12' keys.
* Implemented the following functions:
	tabify
	untabify
	copy-region-as-kill (bound to `M-w')
* Added support for xterm resizing (SIGWINCH signal).
* Added `text-mode-auto-fill' variable that automatically turns on the
  Auto Fill Mode in Text Mode.
* Now `set-variable' suggests the possibile variable values.
* Added `-f function' command line option for calling internal functions
  at Zile startup.
* Bug fixes.

Changes between Zile 1.2 - Zile 1.3

* Implemented region highlighting.
* Added support for non-English character sets.
* Fixed a security hole that made Zile prone to "symlink attack".

Changes between Zile 1.1 - Zile 1.2

* Fixed some memory leaks thanks to the dmalloc package.
* Added alternative key bindings (`M-h' replaces `C-h') that can be set
  using the `alternative-bindings' variable.
* Now the users with home "/" and users not recorded in the `/etc/passwd'
  file are handled correctly.

Changes between Zile 1.0a5 - Zile 1.1

* Fixed a lot of annoying bugs and updated the documentation.
* Added Auto Fill Mode.
* Added C++ Mode (ISO C++ 14882:1998).
* Updated C Mode keywords to latest C standard (ISO C 9899:1999).
* Implemented `goto-line' (M-g).  Contributed by
  Carsten Leonhardt <leo@debian.org>.
* Added backup directory feature.  Contributed by
  Tim Middelkoop <mtim@arranvale.com>.
* Added `describe-function' (C-h d), `describe-variable' (C-h v) and
  `describe-key' (C-h k).  Now the program is really self-documenting!
* Added `list-bindings' (C-h l b), `list-functions' (C-h l f) and
  `list-registers' (C-h l r).
* Now the search functions remember the last searched value.
* All the help functions bound to `M-h' are now bound to `C-h' like Emacs.
* Added a lot of minor but useful features.

WARNING: the file backup variable `backup-files' has been renamed to

src/alist.c
/* Doubly-linked lists
   Copyright (c) 2001-2004 Sandro Sigala.
   Copyright (c) 2004 Reuben Thomas.
   All rights reserved.

   This file is part of Zile.

   Zile is free software; you can redistribute it and/or modify it under
   the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2, or (at your option) any later
   version.

   Zile is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
   for more details.

   You should have received a copy of the GNU General Public License
   along with Zile; see the file COPYING.  If not, write to the Free
   Software Foundation, 59 Temple Place - Suite 330, Boston, MA
   02111-1307, USA.  */

/*	$Id: alist.c,v 1.2 2004/11/15 12:36:13 rrt Exp $	*/

#include <assert.h>
#include <stdlib.h>

#include "config.h"
#include "alist.h"
#include "zile.h"
#include "extern.h"

static aentry aentry_new(void *p)
{
	aentry ae;
	ae = (aentry)zmalloc(sizeof *ae);
	ae->p = p;
	ae->prev = ae->next = NULL;
	return ae;
}

static void aentry_delete(aentry ae)
{
	ae->p = NULL;			/* Useful for debugging. */
	ae->prev = ae->next = NULL;
	free(ae);
}

static aentry find_aentry(alist al, unsigned int i)
{
	aentry ae;
	unsigned int count;
	for (ae = al->head, count = 0; ae != NULL; ae = ae->next, ++count)
		if (count == i)
			return ae;
	return NULL;
}

alist alist_new(void)
{
	alist al;
	al = (alist)zmalloc(sizeof *al);
	al->head = al->tail = al->current = NULL;
	al->idx = 0;
	al->size = 0;
	return al;
}

void alist_delete(alist al)
{
	assert(al != NULL);
	alist_clear(al);
	free(al);
}

void alist_clear(alist al)
{
	aentry ae, next;
	assert(al != NULL);
	for (ae = al->head; ae != NULL; ae = next) {
		next = ae->next;
		aentry_delete(ae);
	}
	al->head = al->tail = al->current = NULL;
	al->idx = 0;
	al->size = 0;
}

void *alist_first(alist al)
{
	assert(al != NULL);
	al->current = al->head;
	al->idx = 0;
	if (al->current != NULL)
		return al->current->p;
	return NULL;
}

void *alist_last(alist al)
{
	assert(al != NULL);
	al->current = al->tail;
	if (al->current != NULL) {
		al->idx = al->size - 1;
		return al->current->p;
	}
	return NULL;
}

void *alist_prev(alist al)
{
	assert(al != NULL);
	if (al->current != NULL) {
		al->current = al->current->prev;
		if (al->current != NULL) {
			--al->idx;
			return al->current->p;
		} else
			return NULL;
	}
	return NULL;
}

void *alist_next(alist al)
{
	assert(al != NULL);
	if (al->current != NULL) {
		al->current = al->current->next;
		if (al->current != NULL) {
			++al->idx;
			return al->current->p;
		} else
			return NULL;
	}
	return NULL;
}

void alist_insert(alist al, unsigned int i, void *p)
{
        assert(i <= al->size);
	if (i == al->size) {
		i = al->size;
		alist_append(al, p);
	} else if (i == 0)
		alist_prepend(al, p);
	else {
		aentry oldae = find_aentry(al, i);
		aentry ae = aentry_new(p);
		ae->next = oldae;
		if (oldae->prev != NULL) {
			ae->prev = oldae->prev;
			ae->prev->next = ae;
		}
		oldae->prev = ae;
		al->current = ae;
		al->idx = i;
		++al->size;
	}
}

void alist_prepend(alist al, void *p)
{
	aentry ae;
	assert(al != NULL);
	ae = aentry_new(p);
	if (al->head != NULL) {
		al->head->prev = ae;
		ae->next = al->head;
		al->head = ae;
	} else
		al->head = al->tail = ae;
	al->current = ae;
	al->idx = 0;
	++al->size;
}

void alist_append(alist al, void *p)
{
	aentry ae;
	assert(al != NULL);
	ae = aentry_new(p);
	if (al->tail != NULL) {
		al->tail->next = ae;
		ae->prev = al->tail;
		al->tail = ae;
	} else
		al->head = al->tail = ae;
	al->current = ae;
	++al->size;
	al->idx = al->size - 1;
}

static void take_off(alist al, aentry ae)
{
	if (ae->prev != NULL)
		ae->prev->next = ae->next;
	if (ae->next != NULL)
		ae->next->prev = ae->prev;
	if (al->head == ae)
		al->head = ae->next;
	if (al->tail == ae)
		al->tail = ae->prev;
	al->current = NULL;
	al->idx = 0;
	--al->size;
}

void alist_remove(alist al)
{
	aentry ae;
	assert(al != NULL);
	if ((ae = al->current) == NULL)
		return;
	take_off(al, ae);
	aentry_delete(ae);
}

void alist_sort(alist al, int (*cmp)(const void *p1, const void *p2))
{
	void **vec;
	aentry ae;
	int i;
	assert(al != NULL && cmp != NULL);
	if (al->size == 0)
		return;
	vec = (void **)zmalloc(sizeof(void *) * al->size);
	for (ae = al->head, i = 0; ae != NULL; ae = ae->next, ++i)
		vec[i] = ae->p;
	qsort(vec, al->size, sizeof(void *), cmp);
	for (ae = al->head, i = 0; ae != NULL; ae = ae->next, ++i)
		ae->p = vec[i];
	free(vec);
}

void *alist_at(alist al, unsigned int i)
{
	aentry ae;
	assert(al != NULL);
	if (i >= al->size)
		return NULL;
	ae = find_aentry(al, i);
	return ae->p;
}

#ifdef TEST

#include <stdio.h>
#include <string.h>

static int sorter(const void *p1, const void *p2)
{
	return strcmp(*(char **)p1, *(char **)p2);
}

/*
 * Stub to make zmalloc &c. happy.
 */
void zile_exit(int exitcode)
{
        exit(exitcode);
}

int main(void)
{
	alist al;
	char *t1 = "def", *t2 = "abc", *t3 = "xyz";
	char *s;

	al = alist_new();
	assert(alist_count(al) == 0);
	assert(alist_current(al) == NULL);
	assert(alist_current_idx(al) == -1);

	alist_append(al, t1);
	assert(alist_count(al) == 1);
	assert(alist_current(al) == t1);
	assert(alist_current_idx(al) == 0);

	alist_append(al, t2);
	assert(alist_count(al) == 2);
	assert(alist_current(al) == t2);
	assert(alist_current_idx(al) == 1);

	s = alist_first(al);
	assert(s == t1);
	assert(alist_current(al) == t1);
	assert(alist_current_idx(al) == 0);

	s = alist_next(al);
	assert(s == t2);
	assert(alist_current(al) == t2);
	assert(alist_current_idx(al) == 1);

	s = alist_next(al);
	assert(s == NULL);
	assert(alist_current(al) == NULL);
	assert(alist_current_idx(al) == -1);

	alist_prepend(al, t3);
	assert(alist_count(al) == 3);
	assert(alist_current(al) == t3);
	assert(alist_current_idx(al) == 0);

	printf("elements:\n");
	for (s = alist_first(al); s != NULL; s = alist_next(al))
		printf("element %d: %s\n", alist_current_idx(al), s);

	alist_sort(al, sorter);
	printf("sorted elements:\n");
	for (s = alist_first(al); s != NULL; s = alist_next(al))
		printf("element %d: %s\n", alist_current_idx(al), s);
        assert(alist_at(al, 0) == t2);
	assert(alist_at(al, 1) == t1);
	assert(alist_at(al, 2) == t3);

	alist_clear(al);
	assert(alist_count(al) == 0);
	assert(alist_current(al) == NULL);
	assert(alist_current_idx(al) == -1);

	alist_insert(al, 0, t1);
	assert(alist_count(al) == 1);
	assert(alist_current(al) == t1);
	assert(alist_current_idx(al) == 0);

	alist_insert(al, 0, t2);
	assert(alist_count(al) == 2);
	assert(alist_current(al) == t2);
	assert(alist_current_idx(al) == 0);

	alist_insert(al, 1, t3);
	assert(alist_count(al) == 3);
	assert(alist_at(al, 0) == t2);
	assert(alist_at(al, 1) == t3);
	assert(alist_at(al, 2) == t1);
	assert(alist_current(al) == t3);
	assert(alist_current_idx(al) == 1);

	alist_delete(al);
	printf("alist test successful.\n");

	return 0;
}

#endif /* TEST */
